Issue/5875 handle inbox note action

JorgeMucientes opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #5875

[changes](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2339) are merged and the FluxC dependency changeset is updated so this is ready for merging. 

### Description
In this PR you'll see changes needed to handle actions over Inbox notes: 

- Handle on click over inbox note actions. Basically, opening an external URL
- Marking the inbox note as actioned in the backend and local DB
- Adding "Dismiss" button to all notes (if they are missing it)
- Handling "Dismiss" clicks. 

### Testing instructions

This feature is hidden under a feature flag for debug builds only, so the generated apk in this PR won't display the "Inbox" option inside More Menu. You'll need to checkout this branch and compile the app to test it. Sorry about that. 

Open Inbox screen and click on different note actions and check: 

- Dismiss button removes the note from the list. In following iteration I'll improve the experience by displaying some progress on the dismiss button while the "Dismiss" api request is executed. 
- Any other action button will open chrome and a URL. This will mark the note as "Actioned" on the API. Marking the note as actioned will remove the "Bold" font style from the title of the note. Beware there is a known issue with the API here. For some reason the API does not mark as "actioned" all notes. So, there are some notes that even if you have clicked on them, they will still display a bold title. 


### Images/gif
<!-- Include before and after images or gifs when appropriate. -->

https://user-images.githubusercontent.com/2663464/162905017-0f2883a4-fcc2-4796-8a56-168b5bb44069.mp4

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

JorgeMucientes committed:
Navigate to url when clicking on inbox note action

JorgeMucientes committed:
Mark note as actioned when action clicked

JorgeMucientes committed:
Add dismiss action to all notes

JorgeMucientes committed:
Handle dismiss action

JorgeMucientes committed:
Merge branch 'issue/5870-save-inbox-notes' into issue/5875-handle-inbox-note-action

JorgeMucientes committed:
Update FluxC version for testing purposes

JorgeMucientes commented:
Ok! don't review this just yet. I'll let you know when is ready for review. Need to add some refactoring in the the base branch of this PR that will then add some changes to this. 

Will let you know when is ready for review again!

JorgeMucientes committed:
Merge branch 'trunk' into issue/5875-handle-inbox-note-action

# Conflicts:
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/inbox/InboxViewModel.kt
#	build.gradle

JorgeMucientes committed:
Update gradle version

JorgeMucientes committed:
Refactor use cases into Repository

JorgeMucientes committed:
Remove unused use cases after adding repositoy

JorgeMucientes committed:
Return Result error from repository functions

JorgeMucientes committed:
Renamed inbox notes status to match code style

JorgeMucientes committed:
Mark inbox note as read visually by removing bold font style

JorgeMucientes committed:
Undo accidentally committed change

nbradbury commented:
@JorgeMucientes Just a heads up that I have a PR in the works which updates the FluxC hasah, but it's failing to compile because the FluxC changes for this PR were merged.

<img width="764" alt="trunk" src="https://user-images.githubusercontent.com/3903757/162819277-143b3837-aa1f-4e85-84b7-b20370566a07.png">

JorgeMucientes committed:
Move repository to domain and remove unneded ObserveInboxNotes use case

JorgeMucientes committed:
Refactor viewModel to have a mutable liveData

JorgeMucientes committed:
Add tests for InboxViewModel

JorgeMucientes committed:
Make fetchInboxNotes return result

JorgeMucientes committed:
Move fromHtml logic out of InboxViewModel

JorgeMucientes committed:
Fix checkstyle issues

JorgeMucientes commented:
Thanks for the heads up on that error @nbradbury! Will try to get this PR merged asap. 

@hafizrahman @atorresveiga this is ready for reviewing, so would any of you be able to take a look at this today?  🙏 
If not, please let me know, so I can request someone else to check this and try to get this merged today 🙂

atorresveiga commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/inbox/InboxViewModel.kt:
```diff
@@ -29,61 +28,108 @@ import javax.inject.Inject
+
+    init {
+        viewModelScope.launch {
+            combine(
```
This is nitpicking but I'm wondering if something like 
```
   init {
        _inboxState.value = InboxState(isLoading = true)
        viewModelScope.launch {
            inboxRepository.fetchInboxNotes()
            inboxNotesLocalUpdates().collectLatest { _inboxState.value = it }
        }
    }
```
is easy to read as loadInboxNotes() is just for initializing _inboxState to `isLoading = true` and calling `inboxRepository.fetchInboxNotes()`

atorresveiga replied:
This is nitpicking but I'm wondering if something like 
```
   init {
        _inboxState.value = InboxState(isLoading = true)
        viewModelScope.launch {
            inboxRepository.fetchInboxNotes()
            inboxNotesLocalUpdates().collectLatest { _inboxState.value = it }
        }
    }
```
is easy to read as loadInboxNotes() is just for initializing _inboxState to `isLoading = true` and calling `inboxRepository.fetchInboxNotes()`

atorresveiga reviewed (approved):
https://user-images.githubusercontent.com/18119390/163048382-ad24621c-e3cb-44cc-9e29-393a39f3cfe4.mp4

I noticed that when opening an external URL and returning to the list the list order changed. Note in the video how the first and second notifications switched places. 
To reproduce this search for 2 notifications with the same date and open/browse the notification, return to the list and check the order the notifications are displayed.
I believe adding a key to the LazyList will fix the problem, something like:
```
 itemsIndexed(
            items = notes,
            key = { index, item: InboxNoteUi -> item.id }
        )
```
but I'm not sure if the item.id is unique

Other than that, it seems good to me and, as this is a minor issue I will pre-approve the PR.
Great job Jorge!👏

atorresveiga commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/inbox/InboxViewModelTest.kt:
```diff
@@ -0,0 +1,141 @@
+import org.mockito.kotlin.whenever
+
+@ExperimentalCoroutinesApi
+class InboxViewModelTest : BaseUnitTest() {
```
Nice! Great job with the tests 👍

JorgeMucientes committed:
Simplify loading logic for Inbox screen

JorgeMucientes commented:
Thanks for checking this out @atorresveiga! Great feedback. 

- I've applied the simplification for loading data on Inbox when entering the screen. Nice suggestion. I initially had only a single flow field to reflect the state of the `InboxViewModel` a that is why I was using `combine`. But then I introduced a liveData backing field to simplify my life when applying some other UI updates and forgot to look into that `init` block and try to make it simpler. So, thanks for that. 

- Regarding the bug on Notes changing order. Good catch! I realised about this yesterday while testing another branch I'm working on. The problem is not the Compose `lazyColumn` so the suggested fix: 
```kotlin
 itemsIndexed(
            items = notes,
            key = { index, item: InboxNoteUi -> item.id }
        )
```
Won't work. Even if the `item.id` is unique. The problem lies on the Room DB implementation in FluxC. When the url of a inbox note is open, behind scenes this inbox note is marked as "actioned". This updates that note in the DB and triggers an update of the Inbox notes list. Right now the query for getting Inbox notes from the DB relies only on the date for ordering:
```kotlin
    @Transaction
    @Query("SELECT * FROM InboxNotes WHERE siteId = :siteId ORDER BY dateCreated DESC")
    abstract fun observeInboxNotes(siteId: Long): Flow<List<InboxNoteWithActions>>
```

 And as you have noticed, there are some notes that have the same date. So the order for notes with the same date might change with every event triggered. I have a fix for this that I'll add in another branch. I introduced the `remoteId` in the ordering sql query. If the dates are the same then notes will be ordered by id:
 ```kotlin
    @Transaction
    @Query("SELECT * FROM InboxNotes WHERE siteId = :siteId ORDER BY dateCreated DESC, remoteId DESC")
    abstract fun observeInboxNotes(siteId: Long): Flow<List<InboxNoteWithActions>>
```

So I will proceed with merging this. And apply the fix in a different PR were I'm adding some FluxC changes. Thanks for testing this thoroughly !!
